pkg/cloudprovider/provider/baremetal/plugins/tinkerbell/client/hardware.go

```diff
@@ -18,94 +18,80 @@ package client
+	hardware := &tinkv1alpha1.Hardware{}
+	if err := h.TinkerbellClient.Get(ctx, client.ObjectKey{Namespace: hardwareRef.Namespace, Name: hardwareRef.Name}, hardware); err != nil {
+		return nil, fmt.Errorf("failed to get hardware '%s' in namespace '%s': %w", hardwareRef.Name, hardwareRef.Namespace, err)
 	}
 
-	if h.GetId() == "" {
-		h.Id = uuid.New().String()
+	return hardware, nil
+}
+
+// SetHardwareID sets the ID of a specified Hardware object.
+func (h *HardwareClient) SetHardwareID(ctx context.Context, hardware *tinkv1alpha1.Hardware, newID string) error {
+	if hardware.Spec.Metadata == nil {
+		hardware.Spec.Metadata = &tinkv1alpha1.HardwareMetadata{}
 	}
 
-	if _, err := t.client.Push(ctx, &hardware.PushRequest{Data: h}); err != nil {
-		return fmt.Errorf("creating hardware in Tinkerbell: %w", err)
+	if hardware.Spec.Metadata.Instance == nil {
+		hardware.Spec.Metadata.Instance = &tinkv1alpha1.MetadataInstance{}
 	}
 
-	return nil
-}
+	hardware.Spec.Metadata.Instance.ID = newID
+	// Set the new ID
+	hardware.Spec.Metadata.State = tbtypes.Staged
+	if newID == "" {
+		// Machine has been deprovisioned
+		hardware.Spec.Metadata.State = tbtypes.Decommissioned
+	}
 
-// Update Tinkerbell Hardware.
-func (t *Hardware) Update(ctx context.Context, h *hardware.Hardware) error {
-	if _, err := t.client.Push(ctx, &hardware.PushRequest{Data: h}); err != nil {
-		return fmt.Errorf("updating template in Tinkerbell: %w", err)
+	// Update the hardware object in the cluster
+	if err := h.TinkerbellClient.Update(ctx, hardware); err != nil {
+		return fmt.Errorf("failed to update hardware ID for '%s': %w", hardware.Name, err)
 	}
 
 	return nil
 }
 
-// Get returns a Tinkerbell Hardware.
-func (t *Hardware) Get(ctx context.Context, id, ip, mac string) (*hardware.Hardware, error) {
-	var method func(context.Context, *hardware.GetRequest, ...grpc.CallOption) (*hardware.Hardware, error)
-
-	req := &hardware.GetRequest{}
-
-	switch {
-	case id != "":
-		req.Id = id
-		method = t.client.ByID
-	case mac != "":
-		req.Mac = mac
-		method = t.client.ByMAC
-	case ip != "":
-		req.Ip = ip
-		method = t.client.ByIP
-	default:
-		return nil, errors.New("need to specify either id, ip, or mac")
+func (h *HardwareClient) GetHardwareWithID(ctx context.Context, uid string) (*tinkv1alpha1.Hardware, error) {
+	// List all hardware in the cluster
+	var hardwares tinkv1alpha1.HardwareList
+	if err := h.TinkerbellClient.List(ctx, &hardwares); err != nil {
+		return nil, fmt.Errorf("failed to list hardware: %w", err)
 	}
 
-	tinkHardware, err := method(ctx, req)
-	if err != nil {
-		if err.Error() == sqlErrorString || err.Error() == sqlErrorStringAlt {
-			return nil, fmt.Errorf("hardware %w", ErrNotFound)
+	// Find the Hardware with the given ID
+	var targetHardware tinkv1alpha1.Hardware
+	for _, hw := range hardwares.Items {
+		if hw.Spec.Metadata.Instance.ID == uid {
+			targetHardware = hw
+			return &targetHardware, nil
```

Comment:
Wouldn't `return &hw, nil` work just as well? Why a new variable if the loop ends due to the return statement anyway?

Reply:
That's true, I guess this left by mistake after cleaning up some code. Will take care of that 😉